app/controllers/volunteers_controller.rb

```diff
@@ -32,9 +32,9 @@ def assign
   end
 
   def shiftless
-    @volunteers = Volunteer.all.keep_if do |volunteer|
-      ((volunteer.region_ids & current_volunteer.region_ids).length > 0) && volunteer.schedule_chains.length == 0
-    end
+    @volunteers = Volunteer.joins(:regions)
+                           .where('regions.id' => current_volunteer.region_ids)
+                           .where('(SELECT COUNT(*) FROM schedule_volunteers WHERE volunteer_id = volunteers.id) = 0')
```

Comment:
Hmmm ok this looks good based on the code / my limited ability to parse SQL.

My concern is: 
New Version finds: 734 volunteers
Old Version finds: 841 volunteers.

Thoughts?

Reply:
Can you give me a count for this query? I think I should have been counting the schedule chains differently:

```
Volunteer.joins(:regions)
  .where('regions.id' => current_volunteer.region_ids)
  .where('(SELECT COUNT(*) FROM schedule_chains ' \
    'INNER JOIN schedule_volunteers ON ' \
    'schedule_chains.id = schedule_volunteers.schedule_chain_id ' \
    "WHERE schedule_chains.active = 't' " \
    'AND schedule_volunteers.volunteer_id = volunteers.id ' \
    "AND schedule_volunteers.active = 't') = 0")
```
🤢

Reply:
841!